data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/router/Router.java

```diff
@@ -24,11 +24,13 @@ public class Router {
     private final RouteEventEvaluator routeEventEvaluator;
     private final DataFlowComponentRouter dataFlowComponentRouter;
     private final Consumer<Event> noRouteHandler;
+    private Set<Record> recordsUnRouted;
```

Comment:
This is not thread-safe. Is there a particular reason you moved it to the field? I say we just keep this as a local variable.

Reply:
This should help make it final:

```
final Set<Record> recordsUnRouted;
        if (!allRecordsRouted) {
            recordsUnRouted = new HashSet<>(allRecords);
        } else {
            recordsUnRouted = null;
        }
```